### PR TITLE
For #10108 - Reconcile list and adapter positions in ExpandableLayout

### PR DESCRIPTION
**Pull request by Mugurell**
For #10108

ExpandableLayout uses adapter based indexes to identify children.
In dynamic menus with list items being hidden this could lead to invalid
calculations and even to exceptions for indexes bigger that the then current
children count in the list.

To reconcile this we'll calculate a valid lastVisibleItemIndexWhenCollapsed and
stickyItemIndex to be used before the operations in calculateCollapsedHeight().

Because the code is now clearly tightly coupled to RecyclerView this patch also
refactors a previous general ViewGroup type to RecyclerView and because of this
part of the previous tests needed some adjustments.

Result in Fenix (observe that upon accessing a subpage of wikipedia the menu contains 2 dynamic options: "Open in app" and "Report Site Issue...". This menu would crash before this patch): 

https://user-images.githubusercontent.com/11428869/115269667-368d2580-a144-11eb-90ed-314e14b3ca9e.mp4

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/browser/menu/src/main/java/mozilla/components/browser/menu/view/ExpandableLayout.kt [hunk 1]
```diff
@@ -360,6 +371,25 @@ internal class ExpandableLayout private constructor(context: Context) : FrameLay
...
+    /**
+     * In a dynamic menu - one in which items or their positions may change the adapter position and
+     * the RecyclerView position for the same item may differ.
+     * This method helps reconcile that.
+     *
+     * @return the RecyclerView position for the item at the [adapterIndex] in the adapter or
+     * [RecyclerView.NO_POSITION] if there is no child for the indicated adapter position.
+     */
+    @VisibleForTesting
+    internal fun getChildPositionForAdapterIndex(listView: RecyclerView, adapterIndex: Int): Int {
```
Comment (Mugurell) [inline components/browser/menu/src/main/java/mozilla/components/browser/menu/view/ExpandableLayout.kt:383]: Thought about having this functionality in a custom getter for `lastVisibleItemIndexWhenCollapsed` and `stickyItemIndex` but then would loose any way of knowing the real passed in value.

**Review by eliserichards — COMMENTED** [review]
The code looks pretty good to me. Thank you for adding the comments to explain your choices. I am going to do some manual tests 👍

**Review by Amejia481 — APPROVED** [review]
Thanks for the patch, I tested locally and I wasn't able to simulate the crash with the patch.

**Review by eliserichards — APPROVED** [review]
Manual tests look good! Let's get this landed 🙌 🥇

**Pull request merged**